apollo-router/src/query_planner/fetch.rs

```diff
@@ -256,32 +451,95 @@ impl Variables {
+                                        // TODO: not great
+                                        let mut new_value = v.clone();
+                                        if let Some(values) = new_value.as_array_mut() {
+                                            for v in values {
+                                                rewrites::apply_single_rewrite(
+                                                    schema,
+                                                    v,
+                                                    &DataRewrite::KeyRenamer({
+                                                        DataKeyRenamer {
+                                                            path: data_path.clone(),
+                                                            rename_key_to: item
+                                                                .rename_key_to
+                                                                .clone(),
+                                                        }
+                                                    }),
+                                                );
+                                            }
+                                        } else {
+                                            rewrites::apply_single_rewrite(
+                                                schema,
+                                                &mut new_value,
+                                                &DataRewrite::KeyRenamer({
+                                                    DataKeyRenamer {
+                                                        path: data_path,
+                                                        rename_key_to: item.rename_key_to.clone(),
+                                                    }
+                                                }),
+                                            );
+                                        }
+                                        return Some((item.rename_key_to.to_string(), new_value));
+                                    }
+                                }
+                                None
+                            }
+                            DataRewrite::ValueSetter(_) => {
+                                // TODO: Log error? panic? not sure
+                                None
+                            }
+                        }
+                    })
+                    .collect();
+
                 let mut value = execute_selection_set(value, requires, schema, None);
                 if value.as_object().map(|o| !o.is_empty()).unwrap_or(false) {
                     rewrites::apply_rewrites(schema, &mut value, input_rewrites);
                     match values.get_index_of(&value) {
                         Some(index) => {
                             inverted_paths[index].push(path.clone());
+                            named_args.push(hash_map);
```

Comment:
@clenfest could we get a weird interaction with entity deduplication here?
Example: `{ topProducts { reviews { author { name } } } }` with `Product` providing context and `User.name` using that context.
We could get in a case where:
- the context from product 1 is used in users 1 and 2
- the context from product 2 is used in users 2 and 3

initially that would have made an entities query with the representations `[{"typename": "User", "id": "1"}, {"typename": "User", "id": "2"}, {"typename": "User", "id": "2"}, {"typename": "User", "id": "3"}]`
but entity deduplication reduces it to   `[{"typename": "User", "id": "1"}, {"typename": "User", "id": "2"}, {"typename": "User", "id": "3"}]`

but here when using context, we would still need two sub operations, one with the contextual variable from product 1 and representations `[{"typename": "User", "id": "1"}, {"typename": "User", "id": "2"}]`, and the other with the contextual variable from product 2 and representations `[{"typename": "User", "id": "2"}, {"typename": "User", "id": "3"}]`

Reply:
I don't think this is an issue as we're only going to perform the query after deduplication is already done if I'm understanding the setup correctly.